### ISSUE TITLE
Implement newAudioDevice/newAudioRecorder in iOS backend

### ISSUE THREAD

**Issue by andresmanz** [libgdx/libgdx#2687]
I need AudioRecorder and AudioDevice on iOS and would like to try to implement it, at least if nobody is currently working on that. I'm not very familiar with iOS development, so: What technologies should these classes use on iOS/RoboVM?

**Comment by andresmanz** [comment]
Alright, thank you. Finally I get to have a look at this now. I will send a PR if everything goes well.

**Comment by schmoofer** [comment]
Someone was able to implement both recording and audio playback with RoboVM: [Stackoverflow post]
(http://stackoverflow.com/questions/27073735/robovm-implementation-of-recording-demo-using-audioqueue-results-in-no-marshal/27218086?noredirect=1#comment44570798_27218086)

I packaged everything into a sample app and already submitted a pull request to RoboVM: [SpeakHere app](https://www.google.com/url?q=https%3A%2F%2Fgithub.com%2Fschmoofer%2Frobovm-samples%2Ftree%2Fmaster%2FSpeakHere&sa=D&sntz=1&usg=AFQjCNHbY-CINk87GB9dbqeaReLZhGPWFw)

The only thing left to do is find out how to access the actual linear PCM audio data during real-time recording and playback. I posted a question about it in the RoboVM Google group [here](https://groups.google.com/forum/#!topic/robovm/wm17rtpxh-s)

If anyone can help out with that, that'd be most excellent.

**Comment by andresmanz** [comment]
@schmoofer Thank you. This is actually how I implemented it for my project. Basically, recording (using AudioRecord's read method) and playing works fine. But when I stop playing/recording, there are some strange (threading-related?) issues with the piped streams. Also, its behaviour in iOS simulator is a bit different from that on an actual device. Yay, surprise. I will have a deeper look at your SpeakHere sample code.

Now as we implemented this using AudioQueues, I'm not sure if it's ok to use that in libgdx as Mario 
expects the use of OpenAL.

**Comment by schmoofer** [comment]
According to the bindings status page for RoboVM ([link](https://github.com/robovm/robovm/wiki/CocoaTouch-bindings-status)), OpenAL hasn't been implemented yet, so that'd be much more work, I assume.

**Comment by Tom-Ski** [comment]
We bind it ourself, see https://github.com/libgdx/libgdx/tree/master/backends/gdx-backend-robovm/src/com/badlogic/gdx/backends/iosrobovm/objectal for examples.

**Comment by schmoofer** [comment]
Here's someone else's bindings [link](https://github.com/threerings/playn/blob/master/robovm/src/playn/robovm/OpenAL.java) for the PlayN framework using RoboVM.

Any chance this can be ported into Libgdx?

**Comment by Tom-Ski** [comment]
We bind this project which uses OpenAL https://github.com/kstenerud/ObjectAL-for-iPhone/tree/master/ObjectAL/ObjectAL

We just bind what we need when we need it.

**Comment by badlogic** [comment]
@Tom-Ski did you just volunteer?! :D

**Comment by Tom-Ski** [comment]
Sure, I can start if and when the gdx-video JNI build stops making me cry.

**Comment by schmoofer** [comment]
Already mentioned it in issue #1828 but I'll reiterate for good measure. I implemented the bindings [here](https://github.com/schmoofer/RecorderLibrary) which also contains the updated IOS backend audio class. @badlogic just notified us that the libgdx crew is working on merging it, so stay tuned y'all.

**Comment by badlogic** [comment]
ello, i'm part of the libGDX crew, looking into it atm :)

**Comment by schmoofer** [comment]
oops, did not mean to exclude the father of LibGDX. Carry on and thanks!

**Comment by badlogic** [comment]
Hrm, i'm not sure what the best way to integrate this is. I don't want to
add another static lib. I'd be cool if this was all Java side, which i
think should be possible. I'll look into it for an hour or so and see where
it leads. Otherwise someone else will have to take this on :)

**Comment by schmoofer** [comment]
Yeah I looked into using only RoboVM classes but they haven't implemented Audio Units yet which are pretty much the basis of my implementation. Is there at least a way to get my IOSAudio.java class to override the stock libGDX one during IPA creation?

Also, here's a sample I help someone else submit using pure Java classes from RoboVM: (link)[https://github.com/schmoofer/robovm-samples/tree/master/SpeakHere]
It uses AudioQueues, but the downside is the read function returns the audio in bytes instead of floats or shorts. I have no idea how to extract the PCM audio data from the byte buffer. Or that's just where my expertise stops.

**Comment by schmoofer** [comment]
My other alternative is to just re-implement some sort of AudioRecorder interface for each platform to bypass libGDX's IOSAudio class, as hacky as it may seem...

**Comment by badlogic** [comment]
Yeah, i just looked into this. You are right, getting this working in pure
Java will be a bit problematic using the queues.

I'm afraid i don't have the time at the moment to properly implement this.
Your best bet atm would be to use Gdx.audio.newAudioRecorder on all
platforms except iOS. On iOS, instantiate your own implementation which
conforms to AudioRecorder, so your game code doesn't care where the
AudioRecorder instance came from.

**Comment by schmoofer** [comment]
No worries, I'll proceed as suggested. Another day, another time.

**Comment by badlogic** [comment]
Actually, we could use the code in
https://github.com/schmoofer/robovm-samples/blob/master/SpeakHere/src/org/robovm/samples/speakhere/AudioRecord.java
converting from bytes to short is trivial, e.g. pseudo code

```java
byte[] pcmBytes = ... // pcmBytes.length must be a multiple of 2
short[] pcmShort = new short[pcmBytes.length / 2]
for(int i = 0, j = 0; i < pcmBytes.length; i+=2, j++) {
   pcmShort[j] = pcmBytes[i] << 8 | pcmBytes[i+1]; // you may have to
switch things around, and maybe mask the bytes with & 0xff
}
```

**Comment by laurence-myers** [comment]
Now that RoboVM has implemented Audio Queues, Audio Units, and all that good stuff, can this be revisited? I would love to use LibGDX as the base for an iOS-compatible synth, but I can't do that without some sort of PCM streaming from a buffer.

**Comment by mpr90** [comment]
Well, there is this [iOS Tone Generator app](http://www.cocoawithlove.com/2010/10/ios-tone-generator-introduction-to.html) which shows how to use AudioUnits to generate tones on the audio output. I am in the process of porting the Objective C code to Java (so far so good) so I can synthesize music/sounds in my app, and it might be a good basis for implementing the AudioDevice functionality in iOS?

**Comment by yarnee** [comment]
i did also implement some audio apps with ibgdx and tried them just now  on ios. Any news about this thread, i also need the newDevise & newRecorder functionality.

**Comment by amitkot** [comment]
Did anyone happen to revisit this issue? 
Is the best way to record audio on iOS still to expose an interface that returns ```Gdx.audio.newAudioRecorder``` on all platforms except iOS, and provide a separate implementation for iOS?

**Comment by euphonialabs** [comment]
I've successfully implemented this function for iOS and, correct, you will have to implement platform-specific functions. Here's the link to the relevant files: [https://github.com/euphonialabs/RecorderLibrary](https://github.com/euphonialabs/RecorderLibrary)

In your *-ios project, add the java files to the src folder and then add the Recorder.a file to the build/libs/ios  folder. Then add the line "<lib>build/libs/ios/Recorder.a</lib>" in the <libs> block in "robovm.xml". From there, follow the instructions at [https://github.com/libgdx/libgdx/wiki/Interfacing-with-platform-specific-code](https://github.com/libgdx/libgdx/wiki/Interfacing-with-platform-specific-code) to interface with the ios recorder functions.

@badlogic Feel free to integrate this solution into master if adequate

**Comment by simernes** [comment]
#5382 the robovm back end is supposedly implemented around 21 days ago.
